### PR TITLE
DLL injection tainted data DFA

### PR DESCRIPTION
**Pull request by dotpaul**
- Generalized tainted data DFA implementation to support other rules
- Hard coded sinks for DLL injection
- Smoke tests for tainted data DFA DLL injection

### CODE REVIEW

**Comment by dotpaul** [comment]
FYI @timhermann @matt-desai

File: src/Microsoft.NetCore.Analyzers/Core/Security/SourceTriggeredTaintedDataAnalyzerBase.cs [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+    using Analyzer.Utilities.Extensions;
+    using Analyzer.Utilities.FlowAnalysis.Analysis.TaintedDataAnalysis;
+    using Microsoft.CodeAnalysis;
+    using Microsoft.CodeAnalysis.Diagnostics;
+    using Microsoft.CodeAnalysis.Operations;
+
+    /// <summary>
+    /// Base class to aid in implementing tainted data analyzers.
+    /// </summary>
+    public abstract class SourceTriggeredTaintedDataAnalyzerBase : DiagnosticAnalyzer
```
Comment (mavasani) [inline src/Microsoft.NetCore.Analyzers/Core/Security/SourceTriggeredTaintedDataAnalyzerBase.cs]: >SourceTriggeredTaintedDataAnalyzerBase [](start = 26, length = 38)

Presuming this is just moving existing code to a new type and so skipping reviewing it - let me know if that is not the case. Thanks!

Comment (mavasani) [reply src/Microsoft.NetCore.Analyzers/Core/Security/SourceTriggeredTaintedDataAnalyzerBase.cs]: Ah, just noticed you moved to use of TaintedDataConfig here. Seems fine to me.

---
In reply to: [239837489](https://github.com/dotnet/roslyn-analyzers/pull/1910#discussion_r239837489) [](ancestors = 239837489)

File: src/Microsoft.NetCore.Analyzers/UnitTests/Security/ReviewCodeForDllInjectionVulnerabilitiesTests.cs [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+    {
+        string input = Request.Form[""in""];
+        AppDomain.CurrentDomain.ExecuteAssembly(input);
+    }
+}
+            ",
+                GetCSharpResultAt(15, 9, 14, 24, "int AppDomain.ExecuteAssembly(string assemblyFile)", "void WebForm.Page_Load(object sender, EventArgs e)", "NameValueCollection HttpRequest.Form", "void WebForm.Page_Load(object sender, EventArgs e)"));
+        }
+
+
```
Comment (mavasani) [inline src/Microsoft.NetCore.Analyzers/UnitTests/Security/ReviewCodeForDllInjectionVulnerabilitiesTests.cs]: Nit: extra new lines.

File: src/Microsoft.NetCore.Analyzers/UnitTests/Security/ReviewCodeForDllInjectionVulnerabilitiesTests.cs [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+{
+    using System;
+    using Microsoft.CodeAnalysis;
+    using Microsoft.CodeAnalysis.Diagnostics;
+    using Microsoft.CodeAnalysis.Testing;
+    using Microsoft.NetCore.Analyzers.Security;
+    using Test.Utilities;
+    using Xunit;
+
+    public class ReviewCodeForDllInjectionVulnerabilitiesTests : TaintedDataAnalyzerTestBase
```
Comment (mavasani) [inline src/Microsoft.NetCore.Analyzers/UnitTests/Security/ReviewCodeForDllInjectionVulnerabilitiesTests.cs]: >ReviewCodeForDllInjectionVulnerabilitiesTests [](start = 17, length = 45)

Please add Traits attribute with dataflow and security traits. I normally only run unit tests with dataflow trait whenever modifying core flow analysis infrastructure/analyses.

File: src/Microsoft.NetCore.Analyzers/UnitTests/Security/ReviewCodeForDllInjectionVulnerabilitiesTests.cs [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+
+namespace Microsoft.NetCore.Analyzers.Security.UnitTests
+{
+    using System;
+    using Microsoft.CodeAnalysis;
+    using Microsoft.CodeAnalysis.Diagnostics;
+    using Microsoft.CodeAnalysis.Testing;
+    using Microsoft.NetCore.Analyzers.Security;
+    using Test.Utilities;
+    using Xunit;
```
Comment (mavasani) [inline src/Microsoft.NetCore.Analyzers/UnitTests/Security/ReviewCodeForDllInjectionVulnerabilitiesTests.cs]: Nit: Not a big deal but we generally put the usings outside namespace in the repo - I am not going to be pedantic about it if you prefer this style.

File: src/Utilities/FlowAnalysis/Analysis/TaintedDataAnalysis/DllSinks.cs [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+        /// <summary>
+        /// <see cref="SinkInfo"/>s for tainted data Dll sinks.
+        /// </summary>
+        public static ImmutableHashSet<SinkInfo> SinkInfos { get; }
+
+        static DllSinks()
+        {
+            ImmutableHashSet<SinkInfo>.Builder sinkInfosBuilder = ImmutableHashSet.CreateBuilder<SinkInfo>();
+
+            sinkInfosBuilder.AddSink(
```
Comment (mavasani) [inline src/Utilities/FlowAnalysis/Analysis/TaintedDataAnalysis/DllSinks.cs]: >sinkInfosBuilder.AddSink [](start = 12, length = 24)

The more I see this, the more I feel you deserve to put these in a configuration file. Let me stop by your office later today to go over the configuration design.

**Comment by mavasani** [comment]
> // Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license 

Ah thanks, we should probably use an analyzer to enforce this. Will try to see if we can use StyleCop analyzers to help us here.

---
Refers to: src/Utilities/Extensions/IDictionaryExtensions.cs:1 in c2298d6. [](commit_id = c2298d65f5574599985460a5ca9e3e469b79d2ac, deletion_comment = False)

File: src/Microsoft.NetCore.Analyzers/UnitTests/Security/ReviewCodeForDllInjectionVulnerabilitiesTests.cs (hunk 3 above)
Comment (dotpaul) [reply src/Microsoft.NetCore.Analyzers/UnitTests/Security/ReviewCodeForDllInjectionVulnerabilitiesTests.cs]: The base class, `TaintedDataAnalyzerTestBase`, has the `[Trait]` on it, and it Works On My Machine(TM).  Will add one for Security.

File: src/Microsoft.NetCore.Analyzers/UnitTests/Security/ReviewCodeForDllInjectionVulnerabilitiesTests.cs (hunk 4 above)
Comment (dotpaul) [reply src/Microsoft.NetCore.Analyzers/UnitTests/Security/ReviewCodeForDllInjectionVulnerabilitiesTests.cs]: I personally prefer usings outside of namespaces too.  Guess I picked up the habit from a previous project.

File: src/Microsoft.NetCore.Analyzers/UnitTests/Security/ReviewCodeForDllInjectionVulnerabilitiesTests.cs (hunk 3 above)
Comment (dotpaul) [reply src/Microsoft.NetCore.Analyzers/UnitTests/Security/ReviewCodeForDllInjectionVulnerabilitiesTests.cs]: Or...I'll figure out what I want to do for Security later.

**Comment by dotpaul** [comment]
Thank you @mavasani !

**Pull request merged**